### PR TITLE
perlPackages.NetSNMP: add patch to support modern SHA algorithms for authentication

### PR DESCRIPTION
**Pull request by jwillikers**
The aged perlPackages.NetSNMP library does not support modern SHA authentication protocols such as SHA-256 and SHA-512. It only supports the less secure SHA-1 algorithm. It's possible to enable support for these newer protocols with a patch from the OpenBSD ports tree found [here](https://github.com/openbsd/ports/blob/master/net/p5-Net-SNMP/patches/patch-lib_Net_SNMP_Security_USM_pm).

References:

* [OpenBSD p5-Net-SNMP patch](https://github.com/openbsd/ports/blob/master/net/p5-Net-SNMP/patches/patch-lib_Net_SNMP_Security_USM_pm)
* [How to use Perl Net::SNMP v3 authentication with newer SHA (SHA256, SHA512) protocols blog post](https://www.claudiokuenzler.com/blog/1205/how-to-perl-net-snmp-v3-authentication-protocol-support-newer-sha-sha256-sha512)
* [Net-SNMP CPAN bug for updated release](https://rt.cpan.org/Public/Bug/Display.html?id=153907)

Patches used in Linux distributions:

* [perl-Net-SNMP Fedora package](https://src.fedoraproject.org/rpms/perl-Net-SNMP/tree/rawhide)
* [libnet-snmp-perl Debian patches](https://sources.debian.org/patches/libnet-snmp-perl/6.0.1-7/)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by zakame** [comment]
Result of `nixpkgs-review pr 349861` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>fusionInventory</li>
    <li>munin</li>
    <li>ocsinventory-agent</li>
    <li>ocsinventory-agent.devdoc</li>
    <li>perl538Packages.NetSNMP</li>
    <li>perl538Packages.NetSNMP.devdoc</li>
    <li>perl540Packages.NetSNMP</li>
    <li>perl540Packages.NetSNMP.devdoc</li>
    <li>smokeping</li>
  </ul>
</details>

**Comment by zakame** [comment]
Thanks for the patch - checking upstream it appears maintenance is stalled and there are actually a few more suggestions of patches, cf https://rt.cpan.org/Public/Bug/Display.html?id=153907

**Comment by jwillikers** [comment]
> Thanks for the patch - checking upstream it appears maintenance is stalled and there are actually a few more suggestions of patches, cf https://rt.cpan.org/Public/Bug/Display.html?id=153907

That's good to hear. Looking into it, those patches seem to align with the patches applied in the Fedora package, seen [here](https://src.fedoraproject.org/rpms/perl-Net-SNMP/tree/rawhide). Maybe I should apply those patches instead?

**Comment by jwillikers** [comment]
> Hi there!
> 
> The patch seems fine to me, but there are some differences between the one vendored in here and the one from OpenBSD ports.
> 
> Nothing bad I'm sure, but there are `} else {` and ` our $VERSION = v4.0.1;` removed in the OpenBSD one that is not removed in this patch.
> 
> ```
> Index: lib/Net/SNMP/Security/USM.pm			      |	diff --git a/lib/Net/SNMP/Security/USM.pm b/lib/Net/SNMP/Secu
> --- lib/Net/SNMP/Security/USM.pm.orig			      |	index 4fb5742..da82212 100644
> +++ lib/Net/SNMP/Security/USM.pm			      |	--- a/lib/Net/SNMP/Security/USM.pm
> @@ -25,9 +25,10 @@ use Net::SNMP::Message qw(		      |	+++ b/lib/Net/SNMP/Security/USM.pm
> 							      >	@@ -25,8 +25,9 @@ use Net::SNMP::Message qw(
>  							      |	+
> +							      |	
>  our $VERSION = v4.0.1;					      <
>  							      |	+
> -   } else {						      <
>  							      |	+
> +							      |	
> +   } else {						      |	    } else {
> 							      >	-
> ```
> 
> Generated with:
> 
> ```shell
> curl -so /tmp/net-smtp-openbsd.patch https://raw.githubusercontent.com/openbsd/ports/ab6365da3ac861e1c56520acdf388d421b357c17/net/p5-Net-SNMP/patches/patch-lib_Net_SNMP_Security_USM_pm
> curl -so /tmp/net-smtp-nixpkgs-pr349861.patch https://raw.githubusercontent.com/NixOS/nixpkgs/fbcd0a2761e960a01eb6dde36a9a26259059c0f5/pkgs/development/perl-modules/net-snmp-enable-newer-sha-algorithms.patch
> diff --suppress-common-lines --color=always -y /tmp/net-smtp-openbsd.patch /tmp/net-smtp-nixpkgs-pr349861.patch
> ```

@stigtsp Thanks for finding that discrepancy. I must have screwed up the patch somehow when I imported it into Git or exported it from Git. I wish I could get `fetchpatch` to work for the patch, but it was giving me some trouble.

**Comment by jwillikers** [comment]
@stigtsp Okay, I manually doctored the patch to ensure it isn't missing anything this time.

**Comment by stigtsp** [comment]
Looks good to me, maybe add a comment above the patch why it's vendored, including the source URL of the patch?

I can also reproduce problems applying this patch using `fetchpatch` but vendoring works.

**Comment by jwillikers** [comment]
I can do that. Since the modules used change, the Perl build files need patched as well to ensure the dependencies are correct, right?

**Comment by jwillikers** [comment]
> Looks good to me, maybe add a comment above the patch why it's vendored, including the source URL of the patch?
> 
> I can also reproduce problems applying this patch using `fetchpatch` but vendoring works.

@stigtsp I figured out how to use `fetchpatch` for the patch.

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -19213,6 +19213,18 @@ with self; {
       url = "mirror://cpan/authors/id/D/DT/DTOWN/Net-SNMP-v6.0.1.tar.gz";
       hash = "sha256-FMN7wcuz883H1sE+DyeoWfFM3P1epUoEZ6iLwlmwt0E=";
     };
+    patches = [
+      (fetchpatch {
+        name = "enable-newer-sha-algorithms.patch";
+        url = "https://raw.githubusercontent.com/openbsd/ports/refs/heads/master/net/p5-Net-SNMP/patches/patch-lib_Net_SNMP_Security_USM_pm";
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:19219]: The URL here should be pinned to a commit, maybe:
```suggestion
        url = "https://raw.githubusercontent.com/openbsd/ports/ab6365da3ac861e1c56520acdf388d421b357c17/net/p5-Net-SNMP/patches/patch-lib_Net_SNMP_Security_USM_pm";
```

**Review by stigtsp — CHANGES_REQUESTED** [review]
Since the patch is now fetched from github, the vendored one can be removed

**Review by stigtsp — CHANGES_REQUESTED** [review]
The dependencies for Net::SNMP should be updated, `DigestHMAC` is needed for `Digest::HMAC_MD5`. 

The other additions seem to be required by the failing test suite (although I didn't look at the failing tests in detail).

```
propagatedBuildInputs = [ CryptDES CryptRijndael DigestHMAC  DigestSHA1 Socket6 ];
```

File: pkgs/top-level/perl-packages.nix (hunk 1 above)
Comment (jwillikers) [reply pkgs/top-level/perl-packages.nix:19219]: Yes of course, my bad.

Comment (jwillikers) [reply pkgs/top-level/perl-packages.nix]: Fixed.

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -19213,6 +19213,18 @@ with self; {
...
+        hash = "sha256-eDfXyYFFvMDQk8xlE6/w3ZWht4E5oupAOpING8cOyRY=";
+        postFetch = ''
+          substituteInPlace $out \
+            --replace "--- lib/Net/SNMP/Security/USM.pm.orig" "--- a/lib/Net/SNMP/Security/USM.pm" \
+            --replace "+++ lib/Net/SNMP/Security/USM.pm" "+++ b/lib/Net/SNMP/Security/USM.pm"
```
[the remark this answers is not in the document]
Comment (jwillikers) [reply pkgs/top-level/perl-packages.nix]: Done

**Comment by jwillikers** [comment]
> Since the patch is now fetched from github, the vendored one can be removed

I've removed that one now.

**Comment by jwillikers** [comment]
> The dependencies for Net::SNMP should be updated, `DigestHMAC` is needed for `Digest::HMAC_MD5`.
> 
> The other additions seem to be required by the failing test suite (although I didn't look at the failing tests in detail).
> 
> ```
> propagatedBuildInputs = [ CryptDES CryptRijndael DigestHMAC  DigestSHA1 Socket6 ];
> ```

I've added a follow-up patch which corrects the dependencies specified in the project.

**Comment by stigtsp** [comment]
Thanks for the great work! I'm going traveling tomorrow, but will have a look as soon as I can after that.

I'm thinking we might consider looking at getting the tests to pass as well, there was a fork referenced in the RT issue here that could have some interesting patches for that:

https://github.com/michal-josef-spacek/Net-SNMP

What do you think @zakame?

**Comment by stigtsp** [comment]
... also if there are some test for the new algos somewhere, that would also be great so we can make sure this keeps working going forward

**Comment by jwillikers** [comment]
Okay, I've done a couple of things.

1. I fixed the tests.
2. I applied the patches from Fedora's package. These are probably the same as https://github.com/michal-josef-spacek/Net-SNMP since Michal Josef Špaček submitted those patches to the Fedora package. There's just a couple less patches in Fedora then there are commits in the repo. These patches correct the dependencies defined in the project so that we don't need to add our own patches to reflect this These patches also have the benefit of adding more tests.
3. I adapted the patch https://github.com/openbsd/ports/blob/master/net/p5-Net-SNMP/patches/patch-lib_Net_SNMP_Security_USM_pm to work on top of the Fedora patches.
4. I included additional tests for each of the new SHA algorithms in the adapted patch.

I've reworked the commits to reflect these changes. Let me know if I should rework or reorder anything. Thanks!

**Comment by stigtsp** [comment]
This is starting to look pretty good I think, thanks again for the great work on it.

I'd like to review the tests and patches a bit more before merging, will try to get this done tomorrow.

File: pkgs/top-level/perl-packages.nix [hunk 3]
```diff
@@ -19213,7 +19213,40 @@ with self; {
...
+    preCheck =
+      (lib.optionalString stdenv.hostPlatform.isLinux ''
+        export NIX_REDIRECTS=/etc/protocols=${pkgs.iana-etc}/etc/protocols
+        export LD_PRELOAD=${pkgs.libredirect}/lib/libredirect.so
+      '');
+    propagatedBuildInputs = [
+      CryptDES
+      CryptRijndael
+      DigestHMAC
+      DigestSHA
```
[the remark this answers is not in the document]
Comment (jwillikers) [reply pkgs/top-level/perl-packages.nix]: I've dropped `DigestSHA`.

**Comment by zakame** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349861`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 11 packages built:</summary>
  <ul>
    <li>fusionInventory</li>
    <li>munin</li>
    <li>nagiosPlugins.check_nwc_health</li>
    <li>nagiosPlugins.check_ups_health</li>
    <li>ocsinventory-agent</li>
    <li>ocsinventory-agent.devdoc</li>
    <li>perl538Packages.NetSNMP</li>
    <li>perl538Packages.NetSNMP.devdoc</li>
    <li>perl540Packages.NetSNMP</li>
    <li>perl540Packages.NetSNMP.devdoc</li>
    <li>smokeping</li>
  </ul>
</details>

**Review by zakame — APPROVED** [review]
LGTM, thanks for the PR! 🎉

**Comment by stigtsp** [comment]
Having some issues building this (on several machines), seemingly caused by fetchpatch2.

```
error: builder for '/nix/store/7sq13pjbsqc10jifq9fzsbjjpkq6c6sz-Net-SNMP-v6.0.1-Rewrite_from_Digest-SHA1-to-Digest-SHA.patch.drv' failed with exit code 1;
       last 9 log lines:
       > error checking the existence of https://tarballs.nixos.org//sha256-dznhj1Fcy0iBBl92p825InjkNZixR2MURVQ/b9bVjtc=:
       > curl: (77) error setting certificate file: /nix/store/h8qhspg7pml84z8s4ghrhmx6m77rs161-nss-cacert-3.104/etc/ssl/certs/ca-bundle.crt
       >
       > trying https://src.fedoraproject.org/rpms/perl-Net-SNMP/raw/6e1d3e8ff2b9bd38dab48301a9d8b5d81ef3b7fe/f/Net-SNMP-v6.0.1-Rewrite_from_Digest-SHA1-to-Digest-SHA.patch
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (77) error setting certificate file: /nix/store/h8qhspg7pml84z8s4ghrhmx6m77rs161-nss-cacert-3.104/etc/ssl/certs/ca-bundle.crt
       > error: cannot download Net-SNMP-v6.0.1-Rewrite_from_Digest-SHA1-to-Digest-SHA.patch from any mirror
       For full logs, run 'nix log /nix/store/7sq13pjbsqc10jifq9fzsbjjpkq6c6sz-Net-SNMP-v6.0.1-Rewrite_from_Digest-SHA1-to-Digest-SHA.patch.drv'.`
```

Unsure why exactly, but observing that rewriting to using `fetchpatch` (and updating hashes due to different patch normalization presumably) works:

```diff
diff --git a/pkgs/top-level/perl-packages.nix b/pkgs/top-level/perl-packages.nix
index 3b533a10e96b..f7d05cc16c00 100644
--- a/pkgs/top-level/perl-packages.nix
+++ b/pkgs/top-level/perl-packages.nix
@@ -7,7 +7,7 @@

 { config
 , stdenv, lib, buildPackages, pkgs, darwin
-, fetchurl, fetchpatch, fetchpatch2, fetchFromGitHub, fetchFromGitLab
+, fetchurl, fetchpatch, fetchFromGitHub, fetchFromGitLab
 , perl, shortenPerlShebang
 , nixosTests
 }:
@@ -19214,25 +19214,25 @@ with self; {
       hash = "sha256-FMN7wcuz883H1sE+DyeoWfFM3P1epUoEZ6iLwlmwt0E=";
     };
     patches = [
-      (fetchpatch2 {
+      (fetchpatch {
         url = "https://src.fedoraproject.org/rpms/perl-Net-SNMP/raw/6e1d3e8ff2b9bd38dab48301a9d8b5d81ef3b7fe/f/Net-SNMP-v6.0.1-Switch_from_Socket6_to_Socket.patch";
-        hash = "sha256-IpVhqI+dXqzauTkLF0Doulg5U33FxHUhqFTp0jeMtMY=";
+        hash = "sha256-AlzNUkAmNSRIIZXs8v5gdp8n7fMfKyqqgM4VDpO47Bs=";
       })
-      (fetchpatch2 {
+      (fetchpatch {
         url = "https://src.fedoraproject.org/rpms/perl-Net-SNMP/raw/6e1d3e8ff2b9bd38dab48301a9d8b5d81ef3b7fe/f/Net-SNMP-v6.0.1-Simple_rewrite_to_Digest-HMAC-helpers.patch";
-        hash = "sha256-ZXo9w2YLtPmM1SJLvIiLWefw7SwrTFyTo4eX6DG1yfA=";
+        hash = "sha256-UFL1rz8HGsU1V+GypybBLvflzpAvh2DYNWjZbfjguvw=";
       })
-      (fetchpatch2 {
+      (fetchpatch {
         url = "https://src.fedoraproject.org/rpms/perl-Net-SNMP/raw/6e1d3e8ff2b9bd38dab48301a9d8b5d81ef3b7fe/f/Net-SNMP-v6.0.1-Split_usm.t_to_two_parts.patch";
-        hash = "sha256-A2gsD6DIX1aFSVLbSL/1zKSM1xiM6hWBadJJH7f5E8o=";
+        hash = "sha256-/tuPmpehUOnCsHgfSAXieQI3QZqEWr67mwohd7NVD2o=";
       })
-      (fetchpatch2 {
+      (fetchpatch {
         url = "https://src.fedoraproject.org/rpms/perl-Net-SNMP/raw/6e1d3e8ff2b9bd38dab48301a9d8b5d81ef3b7fe/f/Net-SNMP-v6.0.1-Add_tests_for_another_usm_scenarios.patch";
-        hash = "sha256-U7nNuL35l/zdSzx1jgjp1PmLQn3xzzDw9DGnyeydi2E=";
+        hash = "sha256-HDRRJTZD+u/ZHgTIx1fVr4U/0DVH92ELfOMAgY++RQE=";
       })
-      (fetchpatch2 {
+      (fetchpatch {
         url = "https://src.fedoraproject.org/rpms/perl-Net-SNMP/raw/6e1d3e8ff2b9bd38dab48301a9d8b5d81ef3b7fe/f/Net-SNMP-v6.0.1-Rewrite_from_Digest-SHA1-to-Digest-SHA.patch";
-        hash = "sha256-dznhj1Fcy0iBBl92p825InjkNZixR2MURVQ/b9bVjtc=";
+        hash = "sha256-jFz5eehV8MHFdd3Wq5S9wWCDhUr5qk4wAJ1mDRJffzw=";
       })
       ../development/perl-modules/net-snmp-add-sha-algorithms.patch
     ];
```

**Comment by stigtsp** [comment]
@ofborg build perlPackages.NetSNMP

**Review by stigtsp — APPROVED** [review]
Looks good to me


## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349861`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 11 packages built:</summary>
  <ul>
    <li>fusionInventory</li>
    <li>munin</li>
    <li>nagiosPlugins.check_nwc_health</li>
    <li>nagiosPlugins.check_ups_health</li>
    <li>ocsinventory-agent</li>
    <li>ocsinventory-agent.devdoc</li>
    <li>perl538Packages.NetSNMP</li>
    <li>perl538Packages.NetSNMP.devdoc</li>
    <li>perl540Packages.NetSNMP</li>
    <li>perl540Packages.NetSNMP.devdoc</li>
    <li>smokeping</li>
  </ul>
</details>

**Pull request merged**